### PR TITLE
DGJ-40 Added support for Supervisor workflow for telework form

### PR DESCRIPTION
**Pull request by simensma-fresh**
## Summary
This PR includes work needed to support the Telework Form Workflow on the Camunda side. This includes changes to the Telework workflow, support for granting read/write/update permissions to a submissions, and some changes to make the BPM a bit more dev friendly.

It also includes support for specifying an ACL for a submission in Forms.io by an admin user.

### Granting permission
In order to grant permissions to a form, you can now use `org.camunda.bpm.extension.hooks.listeners.AddFormPermissionListener` to do so. You specify the email address of the user to grant permissions to and what kind of permissions to grant (write/read/update). 

This will update a ACL on the forms.io side which controls whether or not the user can view / update the submission.

## Changes
- Added a separate BPM prod Dockerfile
- Changed BPM dev Dockerfile to cache maven dependencies and reduce the time needed to view your changes (changes will now be applied if you do a `docker-compose restart forms-flow-bpm` instead of having to rebuild the image
- Assign the approval task to `managerEmail` in the telework form bpmn
- Added `org.camunda.bpm.extension.hooks.listeners.AddFormPermissionListener` BPM listener to grant form permissions
- Added support for submission ACLs in Forms.io
- Made "Tasks" tab of app available for everyone
- Made current "Tasks" available to use in form builder conditionals

## Screenshots (if applicable)

image
image
image

## Notes
To follow:
- Separate PR with documentation on the `AddFormPermissionListener` + how to use current task in conditional

### CODE REVIEW

File: forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            JsonNode jsonNode = objectMapper.readTree(response.getBody());
+            String submissionId = jsonNode.get("_id").asText();
+            return submissionId;
+        } else {
+            throw new FormioServiceException("Unable to create submission for: "+ formUrl+ ". Message Body: " +
+                    response.getBody());
+        }
+    }
+
+    public String grantSubmissionAccess(String formUrl, String user, List<String> permissions) {
```
Comment (simensma-fresh) [inline forms-flow-bpm/src/main/java/org/camunda/bpm/extension/hooks/services/FormSubmissionService.java:83]: This is the only method that was added / modified in this service

**Review by matbusby-fw — APPROVED** [review]
All looks good to me!

File: forms-flow-forms/src/cache/cache.js [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+const debug = {
+  form: require('debug')('formio:cache:form'),
+  loadForm: require('debug')('formio:cache:loadForm'),
+  loadForms: require('debug')('formio:cache:loadForms'),
+  loadFormByName: require('debug')('formio:cache:loadFormByName'),
+  loadFormByAlias: require('debug')('formio:cache:loadFormByAlias'),
+  loadSubmission: require('debug')('formio:cache:loadSubmission'),
+  loadSubmissions: require('debug')('formio:cache:loadSubmissions'),
+  loadSubForms: require('debug')('formio:cache:loadSubForms'),
+  loadSubmissionAcl: require('debug')('formio:cache:loadSubmissionAcl'),
```
Comment (simensma-fresh) [inline forms-flow-forms/src/cache/cache.js:13]: Added this line

File: forms-flow-forms/src/cache/cache.js [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+    * @param req {Object}
+    *   The Express request object.
+    * @param formId {Object|String}
+    *   The submission form id, as BSON or string.
+    * @param subId {Object|String}
+    *   The submission id, as BSON or string.
+    * @param cb {Function}
+    *   The callback function to invoke after loading the submission.
+    */
+    loadSubmissionAcl(req, formId, subId, cb) {
```
Comment (simensma-fresh) [inline forms-flow-forms/src/cache/cache.js:561]: Added this method

File: forms-flow-forms/src/middleware/middleware.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+module.exports = function(router) {
+  return {
+    alias: require('./alias')(router),
+    params: require('./params')(router),
+    accessHandler: require('./accessHandler')(router),
+    addSubmissionResourceAccess: require('./addSubmissionResourceAccess')(router),
+    bootstrapSubmissionResourceAccess: require('./bootstrapSubmissionResourceAccess')(router),
```
Comment (simensma-fresh) [inline forms-flow-forms/src/middleware/middleware.js:9]: Added the two lines above

File: forms-flow-forms/src/models/models.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  // Define our schemas.
+  const models = hook.alter('models', {
+    action: require('./Action')(router.formio),
+    actionItem: require('./ActionItem')(router.formio),
+    form: require('./Form')(router.formio),
+    role: require('./Role')(router.formio),
+    schema: require('./Schema')(router.formio),
+    submission: require('./Submission')(router.formio),
+    token: require('./Token')(router.formio),
+    submissionAcl: require('./SubmissionACL')(router.formio),
```
Comment (simensma-fresh) [inline forms-flow-forms/src/models/models.js:18]: Added this line

File: forms-flow-forms/src/resources/SubmissionResource.js [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+    handlers.afterGet,
+    router.formio.middleware.filterResourcejsResponse(hiddenFields),
+    router.formio.middleware.filterProtectedFields('get', (req) => router.formio.cache.getCurrentFormId(req)),
+  ];
+  handlers.beforePut = [
+    router.formio.middleware.permissionHandler,
+    router.formio.middleware.submissionApplyPatch,
+    router.formio.middleware.filterMongooseExists({field: 'deleted', isNull: true}),
+    router.formio.middleware.bootstrapEntityOwner,
+    router.formio.middleware.bootstrapSubmissionResourceAccess,
```
Comment (simensma-fresh) [inline forms-flow-forms/src/resources/SubmissionResource.js:44]: Added this line

File: forms-flow-forms/src/resources/SubmissionResource.js [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    router.formio.middleware.bootstrapSubmissionAccess,
+    router.formio.middleware.addSubmissionResourceAccess,
```
Comment (simensma-fresh) [inline forms-flow-forms/src/resources/SubmissionResource.js:46]: Added this line

File: forms-flow-web/src/components/Form/Item/Submission/Item/Edit.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  // Pass along the current task with the given submission
+  // so it can be used for validation purposes.
+  const submissionWithTask = _.merge({}, {
+        data: {
+            task: {
+                assignedToMe: task?.assignee && user?.preffered_username === task?.assignee,
+                ...(task || {})
+            }
+        }
+    }, submission);
```
Comment (simensma-fresh) [inline forms-flow-web/src/components/Form/Item/Submission/Item/Edit.js:64]: Added the above

File: forms-flow-web/src/components/PrivateRoute.jsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  return (
+    <>
+      {isAuth ? (
+        <Suspense fallback={<Loading/>}>
+          <Switch>
+            <Route path="/form" component={Form}/>
+            <Route path="/admin" component={Admin}/>
+            <Route path="/formflow" component={Form}/>
+            <ClientReviewerRoute path="/application" component={Application}/>
+            <ClientReviewerRoute path="/task" component={ServiceFlow}/>
```
Comment (simensma-fresh) [inline forms-flow-web/src/components/PrivateRoute.jsx:66]: Changed this from `ReviewerRoute` to `ClientReviewerRoute`

File: forms-flow-web/src/components/Form/Item/Submission/Item/Edit.js (hunk 8 above)
Comment (iman-jamali-fw) [reply forms-flow-web/src/components/Form/Item/Submission/Item/Edit.js:64]: This merge function is very interesting!

**Review by iman-jamali-fw — APPROVED** [review]
This requires huge effort to implement across multiple components. Well done!

File: forms-flow-web/src/components/PrivateRoute.jsx (hunk 9 above)
Comment (iman-jamali-fw) [reply forms-flow-web/src/components/PrivateRoute.jsx:66]: Makes sense!

**Review by matbusby-fw — APPROVED** [review]
Big one!  Looks good from what I can see!

**Pull request merged**